Add support for CrateDB to LangChain LLM framework

amotl opened the pull request:
## About

Discussing the patch to add support for CrateDB to LangChain, to be submitted upstream. **Do not merge.**

## What's inside

- Support for CrateDB's [`FLOAT_VECTOR`](https://crate.io/docs/crate/reference/en/master/general/ddl/data-types.html#float-vector) / [`KNN_MATCH`](https://crate.io/docs/crate/reference/en/master/general/builtins/scalar-functions.html#scalar-knn-match) functionality through LangChain's [vector store](https://python.langchain.com/docs/modules/data_connection/vectorstores/) subsystem.
- Support for loading documents from CrateDB through LangChain's [document loader](https://python.langchain.com/docs/modules/data_connection/document_loaders/) subsystem.
- Support for managing "chat" history using LangChain's [conversational memory](https://python.langchain.com/docs/modules/memory/) subsystem.

## Documentation

- [integrations/providers/cratedb.mdx](https://github.com/crate-workbench/langchain/blob/cratedb/docs/docs/integrations/providers/cratedb.mdx)

## Notebooks
- [integrations/vectorstores/cratedb.ipynb](https://github.com/crate-workbench/langchain/blob/cratedb/docs/docs/integrations/vectorstores/cratedb.ipynb)
- [integrations/document_loaders/cratedb.ipynb](https://github.com/crate-workbench/langchain/blob/cratedb/docs/docs/integrations/document_loaders/cratedb.ipynb)
- [integrations/document_loaders/sqlalchemy.ipynb](https://github.com/crate-workbench/langchain/blob/cratedb/docs/docs/integrations/document_loaders/sqlalchemy.ipynb)
- [integrations/memory/cratedb_chat_message_history.ipynb](https://github.com/crate-workbench/langchain/blob/cratedb/docs/docs/integrations/memory/cratedb_chat_message_history.ipynb)

## Backlog

- [x] See comments below.
- [ ] Break out SQLDatabaseLoader documentation into separate patch.
- [ ] Review all `FIXME` and `TODO` remarks.
- [ ] Bring documentation up to speed. Have a look at those blueprints:
  - https://github.com/langchain-ai/langchain/issues/15664
  - https://python.langchain.com/docs/integrations/vectorstores/faiss
  - https://github.com/langchain-ai/langchain/pull/15750
  - https://github.com/langchain-ai/langchain/pull/16094
  - https://github.com/langchain-ai/langchain/pull/16268

_/cc @matriv, @seut, @marijaselakovic, @karynzv: You may also want to have a review on it? Thanks!_

amotl committed:
CrateDB vector: Add vector store support

The implementation is based on the `pgvector` adapter, as both PostgreSQL and
CrateDB share similar attributes, and can be wrapped well by using the same
SQLAlchemy layer on top.

amotl committed:
CrateDB vector: Add documentation

amotl committed:
CrateDB loader: Add SQLAlchemy document loader

This will become SQLDatabaseLoader later.

amotl committed:
CrateDB loader: Add document loader support

The implementation is based on the generic SQLAlchemy document loader.

amotl committed:
Community: Generalize `SQLChatMessageHistory` to improve code reusability

amotl committed:
CrateDB memory: Add conversational memory support

The implementation is based on the generic `SQLChatMessageHistory`.

amotl commented on libs/langchain/langchain/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,396 @@
+        ]
+        return docs
+
+    def _query_collection(
+        self,
+        embedding: List[float],
+        k: int = 4,
+        filter: Optional[Dict[str, str]] = None,
+    ) -> List[Any]:
+        """Query the collection."""
+        with self.Session() as session:
+            collection = self.get_collection(session)
+            if not collection:
+                raise ValueError("Collection not found")
+
+            filter_by = self.EmbeddingStore.collection_id == collection.uuid
```
About the question how working with different collections works with LangChain. /cc @ckurze

amotl commented on libs/langchain/langchain/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,396 @@
+                    )
+                )
+                .order_by(sqlalchemy.desc("_score"))
+                .join(
+                    self.CollectionStore,
+                    self.EmbeddingStore.collection_id == self.CollectionStore.uuid,
+                )
```
See above.

amotl commented on libs/langchain/langchain/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,396 @@
+            embedding,
+            metadatas=metadatas,
+            ids=ids,
+            collection_name=collection_name,
```
Re: Filtering by collection: Here, `collection_name` is expected as an input argument to the `from_texts()` method.

amotl committed:
CrateDB vector: Fix usage when only reading, and not storing

When not adding any embeddings upfront, the runtime model factory was
not able to derive the vector dimension size, because the SQLAlchemy
models have not been initialized correctly.

amotl committed:
CrateDB vector: Unable to invoke `add_embeddings` without embeddings

amotl committed:
CrateDB vector: Improve SQLAlchemy model factory

From now on, _all_ instances of SQLAlchemy model types will be created
at runtime through the `ModelFactory` utility.

By using `__table_args__ = {"keep_existing": True}` on the ORM entity
definitions, this seems to work well, even with multiple invocations
of `CrateDBVectorSearch.from_texts()` using different `collection_name`
argument values.

While being at it, this patch also fixes a few linter errors.

amotl committed:
CrateDB vector: Fix cascading deletes

When deleting a collection, also delete its associated embeddings.

amotl committed:
CrateDB vector: Add CrateDBVectorSearchMultiCollection

It is a special adapter which provides similarity search across multiple
collections. It can not be used for indexing documents.

amotl committed:
CrateDB vector: Use SA's `bulk_save_objects` method for inserting embeddings

The performance gains can be substantially.

amotl committed:
CrateDB vector: Improve SQLAlchemy data model query utility functions

The CrateDB adapter works a bit different compared to the pgvector
adapter it is building upon: Because the dimensionality of the vector
field needs to be specified at table creation time, but because it is
also a runtime parameter in LangChain, the table creation needs to be
delayed.

In some cases, the tables do not exist yet, but this is only relevant
for the case when the user requests to pre-delete the collection, using
the `pre_delete_collection` argument. So, do the error handling only
there instead, and _not_ on the generic data model utility functions.

amotl committed:
CrateDB vector: Improve testing when initialized without dimensionality

amotl committed:
CrateDB vector: Test non-deterministic values by using pytest.approx

The test cases can be written substantially more elegant.

amotl committed:
CrateDB vector: Fix initialization of vector dimensionality

amotl committed:
CrateDB: Refactor to `langchain_community`

amotl committed:
CrateDB vector: Adjustments for updates to pgvector adapter

amotl commented on libs/langchain/pyproject.toml:
```diff
@@ -110,6 +110,8 @@ databricks-vectorsearch = {version = "^0.21", optional = true}
 couchbase = {version = "^4.1.9", optional = true}
 dgml-utils = {version = "^0.3.0", optional = true}
 datasets = {version = "^2.15.0", optional = true}
+crate = {version = "^0.34.0", extras=["sqlalchemy"], optional = true}
+cratedb-toolkit = {version = ">=0.0.1", optional = true}
```
Better specify `cratedb-toolkit` using an upper version boundary.

amotl replied:
Downstream version dependencies, for testing 3rd-party software used within the "community" section, are no longer part of LangChain's pyproject.toml files.

amotl commented on libs/langchain/tests/integration_tests/cache/fake_embeddings.py:
```diff
@@ -52,7 +52,11 @@ def embed_documents(self, texts: List[str]) -> List[List[float]]:
     def embed_query(self, text: str) -> List[float]:
         """Return consistent embeddings for the text, if seen before, or a constant
         one if the text is unknown."""
-        return self.embed_documents([text])[0]
+        if text not in self.known_texts:
+            return [float(1.0)] * (self.dimensionality - 1) + [float(0.0)]
+        return [float(1.0)] * (self.dimensionality - 1) + [
+            float(self.known_texts.index(text))
+        ]
```
This spot has been resolved differently. We will need to check if that will be fine. Otherwise, we will need to use that code on behalf of a separate `FakeEmbeddings` fixture.

- https://github.com/langchain-ai/langchain/pull/13654#issuecomment-1839809360
- https://github.com/langchain-ai/langchain/pull/14256

amotl replied:
If this patch still needs those adjustments, they will need to be vendorized.

kneth replied:
Can we create an issue to track the work?

amotl replied:
At this spot, we will need to figure out what to do about it. LangChain authors will probably not permit this change, but at the time of writing, we needed this adjustment to satisfy CrateDB integration tests.

amotl replied:
- GH-28

amotl replied:
The code change was considered not to be needed any longer, so it was removed with 41f646270b.

- https://github.com/crate-workbench/langchain/issues/28#issuecomment-2442609786

amotl reviewed (commented):
Just adding a few more notes. I am looking forward to your review.

amotl commented on libs/community/langchain_community/vectorstores/pgvector.py:
The adjustments to the pgvector adapter probably need to be submitted separately.

- 6d36770d
- 890bac3ff

amotl replied:
Both improvements have been submitted to upstream.

- https://github.com/langchain-ai/langchain/pull/16243
- https://github.com/langchain-ai/langchain/pull/16244

amotl committed:
CrateDB vector: Relax test constraint

amotl committed:
CrateDB loader: SQLAlchemyLoader has been superseded by SQLDatabaseLoader

surister reviewed (approved):
Too big to reasonably test today.. looked round it and didn't see anything weird so I'd say LGTM

amotl commented:
@surister: Within the original post, there is a backlog item

> Bring documentation up to speed. Have a look at those blueprints:

If you want to spend a few cycles here, you may want to have a look and evaluate whether you can contribute by improving the documentation correspondingly. Relevant Jupyter Notebooks for CrateDB are surely not top-notch, or even up-to-speed, yet. This patch will be there for a while and everyone is welcome to contribute corresponding refinements.

In the meanwhile, I will be working on a few relevant patches also needed before upstreaming.
- GH-25

amotl committed:
CrateDB: Migrate from `crate[sqlalchemy]` to `sqlalchemy-cratedb`

The CrateDB SQLAlchemy dialect needs more love, so it was separated from
the DBAPI HTTP driver.

amotl commented on docs/docs/integrations/memory/cratedb_chat_message_history.ipynb:
```diff
@@ -0,0 +1,357 @@
+    "class CustomMessage(Base):\n",
+    "\t__tablename__ = \"custom_message_store\"\n",
+    "\n",
+    "\tid = sa.Column(sa.BigInteger, primary_key=True, default=generate_autoincrement_identifier)\n",
```
This has become deprecated. There is a better way to write that down.
```python
id = sa.Column(sa.BigInteger, primary_key=True, server_default=sa.func.now())
```

amotl committed:
CrateDB: Stop using CrateDB Toolkit

amotl commented on docs/docs/integrations/document_loaders/cratedb.ipynb:
```diff
@@ -0,0 +1,232 @@
+   },
+   "outputs": [],
+   "source": [
+    "#!pip install crash 'langchain[cratedb]'"
```
That's no longer a valid installation incantation. Something like this would be better:

```suggestion
    "#!pip install crash langchain sqlalchemy-cratedb"
```

amotl commented on docs/docs/integrations/memory/cratedb_chat_message_history.ipynb:
```diff
@@ -0,0 +1,357 @@
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "!#pip install 'langchain[cratedb]'"
```
That's no longer a valid installation incantation. Something like this would be better:

```suggestion
    "!#pip install langchain sqlalchemy-cratedb"
```

amotl commented on docs/docs/integrations/document_loaders/sqlalchemy.ipynb:
That documentation may already exist upstream. Please check.

- https://github.com/langchain-ai/langchain/pull/18281

amotl replied:
It is not upstream, but will be contributed per [document_loaders/sql_database.ipynb](https://github.com/crate-workbench/langchain/blob/8b278a8a7551f8baea085a331587525ca499bc6f/docs/docs/integrations/document_loaders/sql_database.ipynb).

amotl commented on docs/docs/integrations/memory/cratedb_chat_message_history.ipynb:
```diff
@@ -0,0 +1,357 @@
+    "\n",
+    "class MessageWithDifferentSessionIdColumn(Base):\n",
+    "\t__tablename__ = \"message_store_different_session_id\"\n",
+    "\tid = sa.Column(sa.BigInteger, primary_key=True, default=generate_autoincrement_identifier)\n",
```
This has become deprecated. There is a better way to write that down.
```python
tid = sa.Column(sa.BigInteger, primary_key=True, server_default=sa.func.now())
```

amotl commented on docs/docs/integrations/vectorstores/cratedb.ipynb:
```diff
@@ -0,0 +1,499 @@
+   "outputs": [],
+   "source": [
+    "# Install required packages: LangChain, OpenAI SDK, and the CrateDB Python driver.\n",
+    "!pip install 'langchain[cratedb,openai]'"
```
That's no longer a valid installation incantation. Something like this would be better:

```suggestion
    "!pip install langchain langchain-openai sqlalchemy-cratedb"
```

amotl commented on docs/docs/integrations/vectorstores/cratedb.ipynb:
```diff
@@ -0,0 +1,499 @@
+    "running on [CrateDB Cloud].\n",
+    "\n",
+    "[CrateDB Cloud]: https://console.cratedb.cloud/\n",
+    "[CrateDB using Docker]: https://crate.io/docs/crate/tutorials/en/latest/basic/index.html#docker"
```
```suggestion
    "[CrateDB using Docker]: https://cratedb.com/docs/guide/install/container/"
```

amotl commented on libs/community/langchain_community/vectorstores/cratedb/sqlalchemy_type.py:
```diff
@@ -0,0 +1,84 @@
+
+class FloatVector(UserDefinedType):
+    """
+    https://crate.io/docs/crate/reference/en/master/general/ddl/data-types.html#float-vector
+    https://crate.io/docs/crate/reference/en/master/general/builtins/scalar-functions.html#scalar-knn-match
```
```suggestion
    https://cratedb.com/docs/crate/reference/en/latest/general/ddl/data-types.html#float-vector
    https://cratedb.com/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#scalar-knn-match
```

amotl commented on docs/docs/integrations/providers/cratedb.mdx:
```diff
@@ -0,0 +1,203 @@
+
+
+[CrateDB]: https://github.com/crate/crate
+[CrateDB Cloud]: https://crate.io/product
```
Here and below.

```suggestion
[CrateDB Cloud]: https://cratedb.com/product
```

amotl commented on docs/docs/integrations/vectorstores/cratedb.ipynb:
```diff
@@ -0,0 +1,499 @@
+    "\n",
+    "[CrateDB]: https://github.com/crate/crate\n",
+    "[Elasticsearch]: https://github.com/elastic/elasticsearch\n",
+    "[`FLOAT_VECTOR`]: https://crate.io/docs/crate/reference/en/master/general/ddl/data-types.html#float-vector\n",
+    "[`KNN_MATCH`]: https://crate.io/docs/crate/reference/en/master/general/builtins/scalar-functions.html#scalar-knn-match\n",
```
```suggestion
    "[`FLOAT_VECTOR`]: https://cratedb.com/docs/crate/reference/en/latest/general/ddl/data-types.html#float-vector\n",
    "[`KNN_MATCH`]: https://cratedb.com/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#scalar-knn-match\n",
```

amotl commented on docs/docs/integrations/vectorstores/cratedb.ipynb:
```diff
@@ -0,0 +1,499 @@
+    "[`KNN_MATCH`]: https://crate.io/docs/crate/reference/en/master/general/builtins/scalar-functions.html#scalar-knn-match\n",
+    "[LangChain with CrateDB]: /docs/extras/integrations/providers/cratedb.html\n",
+    "[Lucene]: https://github.com/apache/lucene\n",
+    "[Python client driver for CrateDB]: https://crate.io/docs/python/"
```
```suggestion
    "[Python client driver for CrateDB]: https://cratedb.com/docs/python/"
```

amotl commented on libs/community/tests/integration_tests/document_loaders/docker-compose/cratedb.yml:
```diff
@@ -0,0 +1,20 @@
+
+services:
+  postgresql:
+    image: crate/crate:nightly
```
```suggestion
    image: crate:latest
```

amotl replied:
If someone might be running software tests using that configuration, sticking with `crate/crate:nightly` might be able to catch regressions early. On the other hand, it is not intended for production use. What do you think is applicable when submitting this to the upstream project?

amotl commented on libs/community/tests/integration_tests/vectorstores/docker-compose/cratedb.yml:
```diff
@@ -0,0 +1,20 @@
+
+services:
+  postgresql:
+    image: crate/crate:nightly
```
```suggestion
    image: crate:latest
```

amotl commented on libs/community/langchain_community/vectorstores/cratedb/sqlalchemy_type.py:
Available per sqlalchemy-cratedb. This file can be removed.

amotl commented on libs/community/langchain_community/vectorstores/cratedb/sqlalchemy_patch.py:
Available per sqlalchemy-cratedb. This file can be removed.

amotl commented on libs/community/langchain_community/vectorstores/cratedb/extended.py:
```diff
@@ -0,0 +1,92 @@
+    Provide functionality for searching multiple collections.
+    It can not be used for indexing documents.
+
+    To use it, you should have the ``crate[sqlalchemy]`` Python package installed.
```
That installation incantation will be deprecated. Something like this would be better:

```suggestion
    To use it, you should have the ``sqlalchemy-cratedb`` Python package installed.
```

amotl commented on docs/docs/integrations/providers/cratedb.mdx:
````diff
@@ -0,0 +1,203 @@
+### Install Client
+
+```bash
+pip install 'crate[sqlalchemy]' 'langchain[openai]' 'crash'
````
That installation incantation will be deprecated. Something like this would be better:

```suggestion
pip install crash langchain langchain-openai sqlalchemy-cratedb
```

amotl reviewed (commented):
A few more suggestions coming from a self-review.

amotl committed:
CrateDB: Stop using local `FloatVector` implementation

amotl committed:
CrateDB: Format code. Satisfy linter and type checker. ruff + mypy

kneth reviewed (commented):
I believe that I am not able to provide a meaningful review without a deeper understanding of LangChain.

Said that, I believe it is important work and it is great to see that we are getting closer to complete it.

amotl committed:
CrateDB: Remove adjustment to ConsistentFakeEmbeddings in langchain-core

amotl committed:
CrateDB: Refactor leftovers from langchain-core to langchain-community

amotl committed:
CrateDB: Remove documentation about SQLDatabaseLoader

Those pages have been submitted to LangChain already.

amotl committed:
CrateDB: Remove leftovers in langchain-core

amotl commented:
> I believe that I am not able to provide a meaningful review without a deeper understanding of LangChain.

At [LangChain and CrateDB](https://github.com/crate/cratedb-examples/tree/main/topic/machine-learning/llm-langchain), we are maintaining a set of educational resources to learn a bit about their dynamic duo. That may spark your interest?

amotl commented on docs/docs/integrations/providers/cratedb.mdx:
This document is the main entrypoint that reflects details about CrateDB on the LangChain documentation. Please review it correspondingly.

/cc @ckurze, @kneth, @geragray, @simonprickett, @proddata

amotl commented:
Hi.

> I believe it is important work and it is great to see that we are getting closer to complete it.

The fundamental implementation (vector store, document loader, chat history) as implemented within this genesis patch has been converged into a dedicated package `langchain-cratedb`.

- **PyPI:** https://pypi.org/project/langchain-cratedb/
- **Repository:** https://github.com/crate/langchain-cratedb/
- **Documentation (local):** https://cratedb.com/docs/guide/integrate/langchain/
- **Documentation (remote):** https://python.langchain.com/docs/integrations/providers/cratedb/
- **Examples:** https://github.com/crate/cratedb-examples/tree/main/topic/machine-learning/llm-langchain

We will follow up on this fundamental implementation and add new features and documentation as we go. This PR and feature branch will be closed/deleted in January 2025.

With kind regards,
Andreas.
